misk/src/main/kotlin/misk/ServiceGraphBuilder.kt

```diff
@@ -0,0 +1,133 @@
+   * of the [service] service's dependents can start. Conversely, the dependents of the service service will
+   * be shut down, followed by all [enhancement] enhancements, and finally the [service] service itself.
+   *
+   * @param service The identifier for the service to be enhanced by [enhancement].
+   * @param enhancement The identifier for the service that depends on [service].
+   */
+  fun enhanceService(service: Key<*>, enhancement: Key<*>) {
+    val enhancementSet = enhancementMap[service]
+    enhancementSet!!.add(enhancement)
+  }
+
+  /**
+   * Validates the Service Graph is a valid DAG, then builds a ServiceManager.
+   *
+   * @return ServiceManager that coordinates all services that were registered with this builder.
+   * @throws IllegalStateException if the graph is not valid.
+   */
+  fun build(): ServiceManager {
+    validateDependencyMap()
+    validateEnhancementMap()
+    linkDependencies()
+    checkCycles()
+    return ServiceManager(serviceMap.values)
+  }
+
+  // Builds CoordinatedService2s from the instructions provided in the dependency and enhancement
+  // maps.
+  private fun linkDependencies() {
+    // For each service, add its dependencies and ensure no dependency cycles.
+    for ((key, service) in serviceMap) {
+      // First get the enhancements for the service and handle their downstream dependencies.
+      val enhancements = enhancementMap[key]?.map { serviceMap[it]!! } ?: listOf()
+      service.addEnhancements(enhancements)
+
+      // Now handle regular dependencies.
+      val dependencies = dependencyMap[key]?.map { serviceMap[it]!! } ?: listOf()
+      service.addDependencies(dependencies)
+    }
+  }
+
+  // Checks that no service in this builder has specified a dependency cycle.
+  private fun checkCycles() {
+    val validityMap = mutableMapOf<CoordinatedService2, CycleValidity>()
+
+    for ((_, service) in serviceMap) {
+      val cycle = service.findCycle(validityMap)
+      if (cycle != null) {
+        throw IllegalStateException("Detected cycle: ${cycle.joinToString(" -> ")}")
+      }
+    }
+  }
+
+  // Checks that each service registered with this builder has its dependencies registered.
+  // (i.e. no one service requires a dependency or enhancement that doesn't exist.)
+  private fun validateDependencyMap() {
+    for ((big, dependents) in dependencyMap.asMap()) {
+      if (serviceMap[big] == null) {
+        val stringBuilder = StringBuilder()
+        for (dependent in dependents) {
+          stringBuilder.append("${serviceMap[dependent]}")
+        }
+        throw IllegalStateException("${stringBuilder.toString()} requires $big but no such service "
+            + "was registered with the builder")
+      }
+    }
+  }
+
+  // Checks that no enhancement has been applied more than once.
+  private fun validateEnhancementMap() {
+    val enhancementList = mutableListOf<Key<*>>()
+    for ((_, set) in enhancementMap.asMap()) {
```

Comment:
`for (set in enhancementMap.asMap().values())`

Reply:
or maybe just
```
enhancementList.addAll(enhancementMap.values())
```
?

Reply:
done in b6b460d. went with `+=` instead of `addAll()`